lib/src/emitter.dart

```diff
@@ -0,0 +1,186 @@
+// found in the LICENSE file.
+
+import 'package:code_builder/code_builder.dart';
+import 'package:dart_style/dart_style.dart';
+
+import 'ast.dart';
+import 'token.dart';
+
+String emit(DxAst ast) {
+  final emitter = new _Emitter();
+  ast.accept(emitter);
+  return new DartFormatter().format(emitter.lib.build().accept(new DartEmitter()).toString());
+}
+
+class _Context<T extends Ast> {
+  _Context(this.node, this.context);
+
+  final T node;
+  final dynamic context;
+}
+
+class _Emitter implements AstVisitor {
+  final lib = new FileBuilder();
+  final List<_Context> _nodeStack = <_Context>[];
+
+  void _push(Ast node, [dynamic context]) => _nodeStack.add(new _Context(node, context));
+
+  _Context<T> _ancestorContext<T extends Ast>(bool predicate(_Context c), Object description) {
+    if (_nodeStack.isEmpty) {
+      throw new _EmitterException('Attempted to find $description in the context, but context stack was empty.');
+    }
+
+    return _nodeStack
+      .reversed
+      .skip(1)
+      .firstWhere(predicate, orElse: () {
+        throw new _EmitterException(
+          '$description not found in the context stack. Current stack:\n'
+          '\n'
+          '${_nodeStack.map((c) => 'node: ${c.node.runtimeType}; context: ${c.context.runtimeType}').join('\n')}'
+        );
+      });
+  }
+
+  _Context _pop() => _nodeStack.removeLast();
+
+  @override
+  void visitDxAst(DxAst node) {
+    _push(node);
+    for (Ast child in node.children) {
+      child.accept(this);
+    }
+    _pop();
+  }
+
+  @override
+  void visitImportBlockAst(ImportBlockAst node) {
+    _push(node);
+    for (ImportAst child in node.children) {
+      child.accept(this);
+    }
+    _pop();
+  }
+
+  @override
+  void visitImportAst(ImportAst node) {
+    _push(node);
+    // First token in any import must be an identifier.
+    final Identifier firstToken = node.pathTokens.first;
+    String importUrl;
+    if (node.pathTokens.length == 1) {
+      // Desugar "foo" => "package:foo/foo.dart".
+      importUrl = 'package:${firstToken.name}/${firstToken.name}.dart';
+    } else if (node.pathTokens[1] == Punctuation.$colon) {
+      // Assume a core library import, e.g. "dart:async".
+      assert(node.pathTokens[0] == new Identifier('dart'));
+      assert(node.pathTokens.length == 3);
+      final Identifier lastToken = node.pathTokens.last;
+      importUrl = 'dart:${lastToken.name}';
+    } else {
+      // Assume a package import with subpaths, e.g. "flutter/material",
+      // which is desugared into "package:flutter/material.dart".
+      final buf = new StringBuffer();
+      final tokenIter = node.pathTokens.iterator;
+      bool isFirst = true;
+      while(tokenIter.moveNext()) {
+        if (!isFirst) {
+          buf.write('/');
+        }
+        final Identifier filePart = tokenIter.current;
+        buf.write(filePart.name);
+        if (tokenIter.moveNext()) {
+          assert(tokenIter.current == Punctuation.$slash);
```

Comment:
A common pattern is `expect(token)`, which either consumes the token if it is present or throws/asserts/returns an error code if it isn't

Reply:
The reason I'm only asserting here is that I thought that this kind of AST validation would take place inside the parser (although currently the parser does no such thing). This assert basically ensures that the parser did the right thing. If it triggers, then we likely have a bug in the parser.